backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/lsm/LiveMigrateDiskCommand.java

```diff
@@ -221,64 +221,72 @@ private Set<Guid> getImageGroupIds() {
-            // in the next polling cycle
-            if (!completeLiveMigration()) {
+            case CLONE_IMAGE_STRUCTURE:
+                updateStage(LiveDiskMigrateStage.VM_REPLICATE_DISK_START);
+                replicateDiskStart();
+                updateStage(LiveDiskMigrateStage.IMAGE_DATA_SYNC_EXEC_START);
+                syncImageData();
+                updateStage(LiveDiskMigrateStage.IMAGE_DATA_SYNC_EXEC_END);
                 return true;
-            }
 
-            updateStage(LiveDiskMigrateStage.SOURCE_IMAGE_DELETION);
-            removeImage(getParameters().getSourceStorageDomainId(), getParameters()
-                    .getImageGroupID(), getParameters().getDestinationImageId(), AuditLogType
-                    .USER_MOVE_IMAGE_GROUP_FAILED_TO_DELETE_SRC_IMAGE);
-            updateStage(LiveDiskMigrateStage.LIVE_MIGRATE_DISK_EXEC_COMPLETED);
-            return true;
-        }
+            case IMAGE_DATA_SYNC_EXEC_END:
+            case VM_REPLICATE_DISK_FINISH:
+                updateStage(LiveDiskMigrateStage.VM_REPLICATE_DISK_FINISH);
 
-        if (isRemoveAutoGeneratedSnapshotRequired()) {
-            // This lock is required to prevent CreateSnapshotForVmCommand from
-            // running at the same time as RemoveSnapshotCommand
-            //
-            // This lock was previously acquired by the parent command, but it was released
-            // at the end of CreateSnapshotForVm command called from the executeCommand() method.
-            // Here the lock is acquired again and will be released when this command (LiveMigrateDisk)
-            // finishes.
-            if (!lockManager.acquireLock(getLock()).isAcquired()) {
-                log.info("Failed to acquire VM lock, will retry on the next polling cycle");
+                // in case live migration is yet to complete (VM under heavy I/O load), we shall retry to pivot
+                // in the next polling cycle
+                if (!completeLiveMigration()) {
+                    return true;
+                }
+
+                updateStage(LiveDiskMigrateStage.SOURCE_IMAGE_DELETION);
+                removeImage(getParameters().getSourceStorageDomainId(), getParameters()
+                        .getImageGroupID(), getParameters().getDestinationImageId(), AuditLogType
+                        .USER_MOVE_IMAGE_GROUP_FAILED_TO_DELETE_SRC_IMAGE);
+                updateStage(LiveDiskMigrateStage.LIVE_MIGRATE_DISK_EXEC_COMPLETED);
                 return true;
-            }
 
-            updateStage(LiveDiskMigrateStage.AUTO_GENERATED_SNAPSHOT_REMOVE_START);
-            CommandContext commandContext =
-                    ExecutionHandler.createInternalJobContext(createStepsContext(StepEnum.MERGE_SNAPSHOTS));
-            removeAutogeneratedSnapshot(commandContext, getActionType(), getParameters());
-            updateStage(LiveDiskMigrateStage.AUTO_GENERATED_SNAPSHOT_REMOVE_END);
+            case LIVE_MIGRATE_DISK_EXEC_COMPLETED:
```

Comment:
what about the `getParameters().getLiveDiskMigrateStage() != LiveDiskMigrateStage.AUTO_GENERATED_SNAPSHOT_REMOVE_END;` part that was in isRemoveAutoGeneratedSnapshotRequired?

Reply:
It was removed in [commit #2](https://github.com/oVirt/ovirt-engine/pull/302/commits/78abbb422bf34719ae61634ee734efb7841e82b2) in this PR without changing logics.
Either me and the IDE have bugs in understanding and simplifying boolean logic conditions or `AUTO_GENERATED_SNAPSHOT_REMOVE_END` was not really checked.
If there was an issue it was there before.

Reply:
I don't think it is better, there are too many preparation commits in this PR

Reply:
`preparation commits` are actually "step by step" improvement phases, to make it easier to understand the changes.

Reply:
but this does not help, because they are out of context when you look at them individually

Reply:
but they have no context and make no sense individually, especially when there are five commits each saying

> as a preparation step for the following refactoring steps.

Reply:
ah ok, didn't notice we check the same thing there (getParameters().getLiveDiskMigrateStage()), right
about the separation for multiple commits - when you review PRs on github, you're unlikely to do it commit-by-commit. it was the case before github improved the "changes since last review" view so I don't see much value in that

Reply:
I must say that I'm greatly confused with "what's the bottom line".
In my opinion the new code (assuming it is equivalent to the old one) is much more readable and maintainable and natural and consistent with other `performNextOperation()` implementations that test for at least 2-3 different enum values.
But I will not insist if the verdict is to abort this PR as no-real-value.
Regarding 5 commits, I don't see an issue there, it gives additional option to review the changes if someone wants to better understand that the code is the same:
source-->A-->B-->C-->D-->E==destination
rather than:
source------->destination
But I can reduce the number of commits to 1 if that helps.

Reply:
like I said, I am for keeping only the dead code change [`78abbb4` (#302)](https://github.com/oVirt/ovirt-engine/pull/302/commits/78abbb422bf34719ae61634ee734efb7841e82b2) 

the reward here is far too small